File: football_manager/commercial.py
"""
החיים המסחריים של שחקן: חסויות, תקשורת וסוכנים.

הרעיון: ההצעות שמגיעות אליך הן פונקציה של מי שאתה — מוניטין, גיל,
שערים, כריזמה תקשורתית והמועדון שאתה משחק בו. שחקן בן 19 בליגה
הלאומית מקבל מוסך מקומי; אותו שחקן בגיל 26 אחרי עונה של 24 שערים
בליגת האלופות מקבל טלפון מנייקי, ולא באותו סכום.

וגם: לא כל הצעה חייבת לנחות בשבוע של משחק.
"""

from __future__ import annotations

import random
from typing import Any, Dict, List, Optional

from . import data as D
from .models import Player, clamp


# ---------------------------------------------------------------------------
# מדד הערך המסחרי
# ---------------------------------------------------------------------------

def marketability(player: Player, club_reputation: float = 40.0) -> float:
    """0-100. כמה מותג היה רוצה את הפנים שלך על שלט חוצות."""
    goals = player.career.goals + player.season.goals * 2
    score = player.reputation * 0.62
    score += player.media_skill * 0.20
    score += min(22.0, goals * 0.16)
    score += club_reputation * 0.12
    if player.age <= 23:
        score += 5.0                      # סיפור של כישרון צעיר נמכר טוב
    elif player.age >= 33:
        score -= 6.0
    if player.has_trait("media_darling"):
        score += 9.0
    if player.has_trait("hothead"):
        score -= 4.0
    if player.has_trait("loyal"):
        score += 2.0
    return clamp(score, 0.0, 100.0)


def open_tiers(market: float) -> List[tuple]:
    """אילו דרגי מותגים בכלל מסתכלים עליך."""
    return [tier for tier in D.SPONSOR_TIERS if market >= tier[2]]


# ---------------------------------------------------------------------------
# הצעת חסות
# ---------------------------------------------------------------------------

def sponsor_offer(player: Player, rng: random.Random,
                  club_reputation: float = 40.0,
                  match_week: bool = False) -> Optional[Dict[str, Any]]:
    """בונה הצעת חסות שמתאימה לשחקן הזה עכשיו. None אם אף אחד לא מתעניין."""
    market = marketability(player, club_reputation)
    tiers = open_tiers(market)
    if not tiers:
        return None

    # ככל שהערך המסחרי גבוה, סביר יותר שיתקשר מותג מהדרג העליון
    weights = []
    for index, tier in enumerate(tiers):
        over_min = max(0.0, market - tier[2]) / 30.0
        weights.append(((index + 1) ** 2) * (0.4 + over_min))
    total = sum(weights)
    roll = rng.random() * total
    chosen = tiers[-1]
    for tier, weight in zip(tiers, weights):
        roll -= weight
        if roll <= 0:
            chosen = tier
            break

    key, tier_he, min_rep, base, media_mult, brands = chosen
    kind_key = rng.choice(list(D.DEAL_KINDS))
    kind = D.DEAL_KINDS[kind_key]

    # התשלום נגזר מהדרג, מהערך המסחרי ומסוג החוזה
    over = max(0.0, market - min_rep) / 40.0
    amount = base * kind["pay"] * (0.75 + over * 1.5) * rng.uniform(0.85, 1.25)
    years = rng.randint(1, 3) if key in ("continental", "global") else 1
    shoot_days = max(1, round(kind["days"] * media_mult))

    return {
        "brand": rng.choice(brands),
        "tier": key,
        "tier_he": tier_he,
        "kind": kind_key,
        "kind_he": kind["name"],
        "amount": int(round(amount / 1000) * 1000),
        "years": years,
        "days": shoot_days,
        "media_gain": kind["media"],
        # רק חלק מההצעות דורשות ימי צילום בשבוע משחק
        "clashes": match_week and rng.random() < 0.35,
        "market": round(market, 1),
    }


def deal_summary(offer: Dict[str, Any]) -> str:
    """תיאור ההצעה בשורה אחת."""
    span = "לשנה" if offer["years"] == 1 else f"ל-{offer['years']} שנים"
    return (f"{offer['brand']} · {offer['kind_he']} · "
            f"₪{offer['amount']:,} {span} · {offer['days']} ימי צילומים")


# ---------------------------------------------------------------------------
# עבודות תקשורת
# ---------------------------------------------------------------------------

def media_offer(player: Player, rng: random.Random) -> Optional[Dict[str, Any]]:
    """הצעת עבודה תקשורתית שמתאימה לכריזמה ולמוניטין הנוכחיים."""
    options = [job for job in D.MEDIA_JOBS
               if player.media_skill >= job[2] and player.reputation >= job[3]]
    if not options:
        return None
    key, name, _, _, base = rng.choice(options)
    factor = 0.7 + (player.reputation / 100.0) * 0.9 + (player.media_skill / 100.0) * 0.6
    return {
        "key": key,
        "name": name,
        "amount": int(round(base * factor * rng.uniform(0.9, 1.2) / 1000) * 1000),
    }


# ---------------------------------------------------------------------------
# סוכנים
# ---------------------------------------------------------------------------

def agent_pitch(player: Player, rng: random.Random,
                clubs: List[Any], current_club: Optional[Any]) -> Optional[Dict[str, Any]]:
    """סוכן שמזהה שאתה שווה יותר ממה שאתה מקבל, ומביא יעד קונקרטי."""
    current_rep = current_club.reputation if current_club else 20
    targets = [c for c in clubs
               if (not current_club or c.cid != current_club.cid)
               and c.reputation > current_rep + 6
               and c.reputation <= player.reputation + 26]
    if not targets:
        return None
    target = rng.choice(targets)
    raise_factor = rng.uniform(1.25, 2.1) * (1 + (target.reputation - current_rep) / 130.0)
    return {
        "agent": rng.choice(D.AGENT_NAMES),
        "club": target.cid,
        "club_name": target.name,
        "wage": int(round(player.contract.wage * raise_factor / 500) * 500),
        "fee": int(round(player.contract.wage * raise_factor * 0.12 / 500) * 500),
    }

File: web/commercial.js
// ---------------------------------------------------------------------------
// החיים המסחריים — פורט מ-football_manager/commercial.py.
// ההצעות שמגיעות אליך הן פונקציה של מי שאתה: מוניטין, שערים, כריזמה
// והמועדון. אותו אירוע נראה אחרת בגיל 18 ובגיל 27.
// ---------------------------------------------------------------------------

/** 0-100. כמה מותג היה רוצה את הפנים שלך על שלט חוצות. */
function marketability(p, clubReputation = 40) {
  const goals = p.career.goals + p.season.goals * 2;
  let score = p.reputation * 0.62;
  score += p.mediaSkill * 0.20;
  score += Math.min(22, goals * 0.16);
  score += clubReputation * 0.12;
  if (p.age <= 23) score += 5;
  else if (p.age >= 33) score -= 6;
  if (hasTrait(p, "media_darling")) score += 9;
  if (hasTrait(p, "hothead")) score -= 4;
  if (hasTrait(p, "loyal")) score += 2;
  return clamp(score, 0, 100);
}

function openTiers(market) {
  return D.SPONSOR_TIERS.filter(tier => market >= tier[2]);
}

/** בונה הצעת חסות שמתאימה לשחקן הזה עכשיו. null אם אף אחד לא מתעניין. */
function sponsorOffer(p, rng, clubReputation = 40, matchWeek = false) {
  const market = marketability(p, clubReputation);
  const tiers = openTiers(market);
  if (!tiers.length) return null;

  const weights = tiers.map((tier, index) =>
    Math.pow(index + 1, 2) * (0.4 + Math.max(0, market - tier[2]) / 30));
  const total = weights.reduce((a, b) => a + b, 0);
  let roll = rng.random() * total;
  let chosen = tiers[tiers.length - 1];
  for (let i = 0; i < tiers.length; i++) {
    roll -= weights[i];
    if (roll <= 0) { chosen = tiers[i]; break; }
  }

  const [tierKey, tierHe, minRep, base, mediaMult, brands] = chosen;
  const kindKey = rng.choice(Object.keys(D.DEAL_KINDS));
  const kind = D.DEAL_KINDS[kindKey];

  const over = Math.max(0, market - minRep) / 40;
  const amount = base * kind.pay * (0.75 + over * 1.5) * rng.uniform(0.85, 1.25);
  const years = (tierKey === "continental" || tierKey === "global") ? rng.randint(1, 3) : 1;

  return {
    brand: rng.choice(brands),
    tier: tierKey,
    tierHe,
    kind: kindKey,
    kindHe: kind.name,
    amount: Math.round(amount / 1000) * 1000,
    years,
    days: Math.max(1, Math.round(kind.days * mediaMult)),
    mediaGain: kind.media,
    clashes: matchWeek && rng.random() < 0.35,
    market: Math.round(market * 10) / 10,
  };
}

function dealSummary(offer) {
  const span = offer.years === 1 ? "לשנה" : `ל-${offer.years} שנים`;
  return `${offer.brand} · ${offer.kindHe} · ₪${fmt(offer.amount)} ${span} · ${offer.days} ימי צילומים`;
}

/** הצעת עבודה תקשורתית שמתאימה לכריזמה ולמוניטין הנוכחיים. */
function mediaOffer(p, rng) {
  const options = D.MEDIA_JOBS.filter(job => p.mediaSkill >= job[2] && p.reputation >= job[3]);
  if (!options.length) return null;
  const [key, name, , , base] = rng.choice(options);
  const factor = 0.7 + (p.reputation / 100) * 0.9 + (p.mediaSkill / 100) * 0.6;
  return { key, name, amount: Math.round(base * factor * rng.uniform(0.9, 1.2) / 1000) * 1000 };
}

/** סוכן שמזהה שאתה שווה יותר ממה שאתה מקבל, ומביא יעד קונקרטי. */
function agentPitch(p, rng, clubs, currentClub) {
  const currentRep = currentClub ? currentClub.reputation : 20;
  const targets = clubs.filter(c =>
    (!currentClub || c.cid !== currentClub.cid)
    && c.reputation > currentRep + 6
    && c.reputation <= p.reputation + 26);
  if (!targets.length) return null;
  const target = rng.choice(targets);
  const raise = rng.uniform(1.25, 2.1) * (1 + (target.reputation - currentRep) / 130);
  return {
    agent: rng.choice(D.AGENT_NAMES),
    club: target.cid,
    clubName: target.name,
    wage: Math.round(p.contract.wage * raise / 500) * 500,
    fee: Math.round(p.contract.wage * raise * 0.12 / 500) * 500,
  };
}
